File: src/server/config/seed.js
const fs = require('fs');
const path = require('path');
const mongoose = require('mongoose');
require('dotenv').config();

// 1. Schema Definition
const propertySchema = new mongoose.Schema({
  propertyId: { type: Number, required: true, unique: true },
  title: { type: String, required: true },
  price: { type: Number, required: true },
  location: { type: String, required: true },
  specs: {
    bedrooms: Number,
    bathrooms: Number,
    sizeSqft: Number,
    amenities: [String],
  },
  media: {
    imageUrl: String,
  },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
});

const Property = mongoose.model('Property', propertySchema);

async function seedDatabase() {
  try {
    const MONGODB_URI = process.env.MONGODB_URI || 'mongodb://localhost:27017/realestate';
    await mongoose.connect(MONGODB_URI);
    console.log('✅ Connected to MongoDB');

    // Paths to raw data
    const basicsPath = path.join(__dirname, '../../../Case Study 3 JSON 1 (1).txt');
    const characteristicsPath = path.join(__dirname, '../../../Case Study 3 JSON 2.txt');
    const imagesPath = path.join(__dirname, '../../../Case Study 3 JSON 3.txt');

    // Load data
    const basics = JSON.parse(fs.readFileSync(basicsPath, 'utf8'));
    const characteristics = JSON.parse(fs.readFileSync(characteristicsPath, 'utf8'));
    const images = JSON.parse(fs.readFileSync(imagesPath, 'utf8'));

    console.log('📦 Loading and merging data...');

    const unifiedProperties = basics.map(b => {
      const char = characteristics.find(c => c.id === b.id);
      const img = images.find(i => i.id === b.id);

      if (!char || !img) {
        throw new Error(`Data inconsistency found for ID ${b.id}. Missing characteristic or image.`);
      }

      return {
        propertyId: b.id,
        title: b.title,
        price: b.price,
        location: b.location,
        specs: {
          bedrooms: char.bedrooms,
          bathrooms: char.bathrooms,
          sizeSqft: char.size_sqft,
          amenities: char.amenities,
        },
        media: {
          imageUrl: img.image_url,
        },
      };
    });

    // Clean existing data to prevent duplicates during seeding
    await Property.deleteMany({});

    // Bulk insert
    await Property.insertMany(unifiedProperties);
    console.log(`🚀 Successfully seeded ${unifiedProperties.length} properties into MongoDB!`);

  } catch (error) {
    console.error('❌ Seeding failed:', error.message);
    process.exit(1);
  } finally {
    await mongoose.connection.close();
  }
}

seedDatabase();

File: src/backend/config/seed.py
import os
import json
from pymongo import MongoClient
from dotenv import load_dotenv

load_dotenv()

def seed_database():
    try:
        MONGODB_URI = os.getenv('MONGODB_URI', 'mongodb://localhost:27017/')
        client = MongoClient(MONGODB_URI)
        db = client['realestate']
        collection = db['properties']

        # Paths to raw data
        basics_path = 'Case Study 3 JSON 1 (1).txt'
        chars_path = 'Case Study 3 JSON 2.txt'
        imgs_path = 'Case Study 3 JSON 3.txt'

        # Load data
        with open(basics_path, 'r') as f: basics = json.load(f)
        with open(chars_path, 'r') as f: characteristics = json.load(f)
        with open(imgs_path, 'r') as f: images = json.load(f)

        print("📦 Loading and merging data...")

        unified_properties = []
        for b in basics:
            char = next((c for c in characteristics if c['id'] == b['id']), None)
            img = next((i for i in images if i['id'] == b['id']), None)

            if not char or not img:
                print(f"⚠️ Data inconsistency found for ID {b['id']}. Skipping...")
                continue

            unified_properties.append({
                "propertyId": b['id'],
                "title": b['title'],
                "price": b['price'],
                "location": b['location'],
                "specs": {
                    "bedrooms": char['bedrooms'],
                    "bathrooms": char['bathrooms'],
                    "sizeSqft": char['size_sqft'],
                    "amenities": char['amenities'],
                },
                "media": {
                    "imageUrl": img['image_url'],
                }
            })

        # Clear and insert
        collection.delete_many({})
        collection.insert_many(unified_properties)
        print(f"🚀 Successfully seeded {len(unified_properties)} properties into MongoDB!")

    except Exception as e:
        print(f"❌ Seeding failed: {e}")

if __name__ == "__main__":
    seed_database()
